### PR TITLE
Add support for Hack

### PR DESCRIPTION
**Pull request by theHamsta**

### CODE REVIEW

File: queries/hack/highlights.scm
Comment (theHamsta) [inline queries/hack/highlights.scm:41]: TODO: `@type.builtin`

**Comment by theHamsta** [comment]
Based on https://github.com/nvim-treesitter/nvim-treesitter/pull/2031 to add tests. Certainly, not complete @4e554c4c could you review and provide annotated test files like the ones in this commit if something is wrong? We can also merge now and you can continue refining the queries.

Maybe it would be nice to have `$` (like for PHP) and `?` in `$foo` and `?Type` as visible tokens in the parser. Would require an upstream change

**Comment by theHamsta** [comment]
Is the filetype `hack` or `hacklang`

**Comment by 4e554c4c** [comment]
@theHamsta I've tried installing the playground and looking at some of my job's code on this branch and it looks great!
I think I've found a couple of parser errors though. These are most likely upstream issues, but I can definitely provide some minimal examples.
I'm having a bit of trouble running the tests, however, even after installing highlight-assertions. For the hacklang tests I'm getting
```
Vim:E474: Unidentified byte: Error: /home/pounce/build/nvim-treesitter/parser/hack.so: cannot open shared object file: No such file or directory
```
Any idea why?

**Comment by 4e554c4c** [comment]
also https://github.com/hhvm/vim-hack/blob/main/ftdetect/hack.vim has a better ftdetect

**Comment by theHamsta** [comment]
@4e554c4c did you run `:TSInstall hack`

**Comment by 4e554c4c** [comment]
LGTM! Before there were issues with type constants and shapes but those seem to be resolved now
test case showing this:
```
<?hh

class C extends Superclass implements Iface {
//      ^ keyword          ^ keyword
  use Trait;
  // < keyword
  const type X = shape(
  // <- keyword ^ type.builtin
    "a" => int,
  // ^ string
    "b" => string,
  //       ^ type.builtin
  );
}
```

File: ftdetect/hack.vim
Comment (theHamsta) [inline ftdetect/hack.vim:12]: for me `BufRead,BufNewFile *.hack,*.hackpartial,*.hck` would be fine

**Comment by theHamsta** [comment]
@stsewd can we merge this?

File: queries/hack/highlights.scm
Comment (vigoux) [inline queries/hack/highlights.scm:36]: I am neatpicky but...
```suggestion
 "insteadof"
```

File: queries/hack/highlights.scm
Comment (vigoux) [inline queries/hack/highlights.scm:65]: I don't know hack, but is this a null type, or the constant `null` ?

File: tests/query/highlights/hack/attribute-type.hack
Comment (vigoux) [inline tests/query/highlights/hack/attribute-type.hack:12]: Why is that ?

**Review by vigoux — COMMENTED** [review]
I am going to trust you for most of this, just left a few comments.

File: queries/hack/highlights.scm
Comment (theHamsta) [reply queries/hack/highlights.scm]: Fixed it to constant.builtin.

File: tests/query/highlights/hack/attribute-type.hack
Comment (theHamsta) [reply tests/query/highlights/hack/attribute-type.hack]: (cannot capture keyword "function" as keyword.function as it does not appear in AST). This has to be fixed upstream. The current workaround is to highlight all of `function foo` as `function`

Comment (theHamsta) [reply tests/query/highlights/hack/attribute-type.hack]: This should be changed to https://github.com/slackhq/tree-sitter-hack/blob/dd7c1bb8b11d290db4aec173740efdb510f0bba9/grammar.js#L455 `alias(/function(s)?/, "function")`

Comment (vigoux) [reply tests/query/highlights/hack/attribute-type.hack]: Might be worth opening a PR then ?

Comment (theHamsta) [reply tests/query/highlights/hack/attribute-type.hack]: https://github.com/slackhq/tree-sitter-hack/pull/36

Comment (theHamsta) [reply tests/query/highlights/hack/attribute-type.hack]: image

This is very creepy: https://github.com/slackhq/tree-sitter-hack/issues/37

Comment (vigoux) [reply tests/query/highlights/hack/attribute-type.hack]: Great, then I would wait for the PR to be merged, so that we directly pull the changes.

Comment (theHamsta) [reply tests/query/highlights/hack/attribute-type.hack]: I don't think the upstream repo is active anymore.

Comment (4e554c4c) [reply tests/query/highlights/hack/attribute-type.hack]: I spoke to some folks at slack and they said they'd look into it, it looks like slackhq/tree-sitter-hack#37 has been fixed

**Comment by theHamsta** [comment]
Outstanding issue here is the merge of https://github.com/slackhq/tree-sitter-hack/pull/36 . But I don't think this should block this PR

File: ftdetect/hack.vim
Comment (clason) [inline ftdetect/hack.vim:14]: I don't like that this mixes `filetype` and `syntax`. I believe the proper thing here is to make `hack` a subtype of `php`, as in `php.hack`?

File: ftdetect/hack.vim
Comment (clason) [inline ftdetect/hack.vim:18]: If it doesn't use the `hack` syntax, it doesn't belong here? Proper filetype support should go upstream (Vim).

Comment (theHamsta) [reply ftdetect/hack.vim:18]: Let's just remove the ftdetect.

Comment (theHamsta) [reply ftdetect/hack.vim:18]: hack is already on the tracking issue.

**Review by clason — CHANGES_REQUESTED** [review]
I'm not familiar enough with Hack and PHP to comment on the queries, but I trust you there.

**Comment by theHamsta** [comment]
@clason me neither, we have to trust https://github.com/nvim-treesitter/nvim-treesitter/pull/2042#issuecomment-992765190

**Review by clason — APPROVED** [review]

File: README.md
Comment (clason) [inline README.md:192]: should this get a checkmark now, that basic highlight queries are included?

Comment (theHamsta) [reply README.md:192]: the checkmark depends on a maintainer being present. It's auto-generated.

Comment (clason) [reply README.md:192]: Ah, I see -- then we probably want to change the README (because that states the checkmark means "basically usable, has queries"; my bad!)

Since the maintainer information is already explicit, I'd prefer to use the checkmark to indicate independent information -- such as basic usability.

Comment (theHamsta) [reply README.md:192]: sure, sounds reasonable.

File: ftdetect/hack.vim
Comment (fredemmott) [reply ftdetect/hack.vim]: fwiw, this is historical: Hack used to be a superset of PHP, so in hhvm/vim-hack, we set filetype=hack to mark what tooling should be used, but we dealt with the Hack syntax by extending the PHP syntax.

It's not a good way to do things nowadays, and I'm happy to see the syntax=php isn't in the final version of this PR :)

That said: `.hhi` should be treated as a hack file, and probably should have the `hack` syntax. They're not used often - they're effectively our equivalent of header files, and used to give static analysis information about the builtin functions that are implemented in C++.

**Pull request merged**
